index.js

```diff
@@ -242,53 +242,78 @@ const parseArgs = (config = { __proto__: null }) => {
         }
       }
       ArrayPrototypeUnshiftApply(remainingArgs, expanded);
+      groupCount = expanded.length;
       continue;
     }
 
     if (isShortOptionAndValue(arg, options)) {
       // e.g. -fFILE
       const shortOption = StringPrototypeCharAt(arg, 1);
-      const longOption = findLongOptionForShort(shortOption, options);
-      const optionValue = StringPrototypeSlice(arg, 2);
-      checkOptionUsage(longOption, optionValue, options, `-${shortOption}`, strict, allowPositionals);
-      storeOption(longOption, optionValue, options, result.values);
+      const optionName = findLongOptionForShort(shortOption, options);
+      const value = StringPrototypeSlice(arg, 2);
+      ArrayPrototypePush(
+        elements,
+        { kind: 'option', optionName, isShort: true, argIndex,
```

Comment:
```suggestion
        { kind: 'option', optionName, short: true, argIndex,
```

Just `short`. The web platform has an [official guideline about this](https://w3ctag.github.io/design-principles/#naming-booleans), which node does not always follow, but generally should.

Reply:
I made that change after writing code which used element.short (boolean) and options[foo].short in the same expression. Could go back to the character I had in a previous iteration! 😄

Reply:
I think the character is actually more misleading, because it implies that it means the same thing as in it does in the config, i.e. "_could be_ short", whereas it actually means "was definitely short in this particular use" - or, in some cases, "was _not_ short in this particular use". So you might see `{ optionName: 'foo', short: undefined }` for an option which actually was configured with `short: 'f'`, which is weird.

I don't personally think the use of the same property name is a big deal, since these aren't really contexts that you're likely to conflate the two. But maybe there's some other name we could go with. E.g., invert it and call it `long`. That seems worse to me than just having a boolean `short`, but it's technically an option.

Reply:
New approach. I swapped `isShort` for `optionUsed`, which has value like `-a` or `--alpha`. (Simple string even when `inlineValue`, or from a short option group.)

1) I need some meta-data to construct error messages. Instead of always needing to reconstruct short-or-long, supply a potentially useable string directly.
2) uses which care about short vs long can easily inspect the string. Indirect but easy.

Reply:
Ehhhhh that still feels needless complex to me, relative to just having a `short` boolean.

Reply:
I think `token.optionUsed` will be commonly used in error messages. With it available, to implement the existing parsing I only needed to deduce short-ness in one place for additional lines in extended message.

Calculating `optionUsed` using `short` is non-trivial. In a library built on `parseArgs` I think this would be robust given an option token:

```
const optionUsed = (!token.short) ?
  `--${token.optionName}` : 
  `-${options?.[token.optionName]?.short ?? token.optionName}`
```

Another calculation we use is for "shortAndLong". I was going to argue this is easier, but I think I spotted a bug for an unrecognised short option! However, this is used for an optional extended format and library author could just use `optionUsed`, so feels less core.

```
const short = options?.[element.optionName]?.short;
const shortAndLong = `${short ? `-${short}, ` : ''}--${token.optionName}`;
```

Reply:
>  In a library built on `parseArgs` I think this would be robust given an option token:

is ``token.short ? `-${options[token.optionName].short}` : `--${token.optionName}` `` not sufficient?

Reply:
Thoughts on `token.raw` or `token.rawOption`? e.g.

<details>
  <summary>acorn</summary>

![Screen Shot 2022-05-31 at 12 28 30 PM](https://user-images.githubusercontent.com/32409546/171270609-7470dfc1-5a7d-4a85-8df5-fab8e3091fd1.png)
</details>

<details>
  <summary>@babel/parser</summary>

![Screen Shot 2022-05-31 at 12 29 23 PM](https://user-images.githubusercontent.com/32409546/171270679-874d0f92-0268-4414-92f5-6a4cc4b1b3b4.png)
</details>

Reply:
The raw flavours only seem useful if the value has been modified, which we don't currently do?

(Thanks for AST examples.)

Reply:
Looks like we're modifying the arguments to me e.g.
```js
// node cli.js -f --bar=baz

const {tokens} = parseArgs()
/*
tokens === [
  {
    optionName: 'foo', // Modified by resolving the `longOption` and removing the `-` prefix
    raw: '-f',
    ...
  },
  {
    optionName: 'bar', // Modified by removing the `--` prefix and extracting the value
    value: 'baz',
    raw: '--bar=baz',
  },
]
*/
```

Reply:
Oh, I see, I had "rawValue" in my head from the babel example.

I had wondered about returning the argument string. It is accessible using `args[token.index]` which I think is probably good enough. With caveat that args might be auto-determined, but author has simple work-around to supply the args.

Unlike a "real" AST, I'm not trying to point back into the "text" itself. (Both your examples had `start` and `end` characters positions.) Two cases where I am unsure what "rawOption" should even be are token for option+space+value, and token for short option group.

Reply:
Pretty sure everything we've discussed can be derived from this minimal set `token` details:

```js
if (token.kind === 'option') {
  const long = token.raw.charAt(1) === '-'
  const short = token.raw.charAt(1) !== '-'
  const inline = token.raw.includes('=')
  const shortOptionGroup = short && token.raw.length > 2

  const shortOption = config.options[token.name]?.short
  const optionUsed = short ? `-${config.options[token.name].short}` : `--${token.name}` 
  const shortAndLong = `${short ? `-${config.options[token.name].short}, ` : ''}--${token.name}`
}
```

Reply:
> Pretty sure everything discussed can be derived from this minimal set of token details:

Maybe, but may require quite deep understanding or subtle code. As I may be able to show...

I think it is appropriate to offer meta-data which is commonly needed by library authors and hard to determine afterwards, and often available during the tokenising. I included `optionUsed` as common for error messages and tricky (and I wanted it or something to make it easier). I haven't needed `inlineValue` yet, but I think it will be commonly used when adding coerce type post-processing.

`const long = token.raw.charAt(1) === '-'`

Agree.

`  const short = token.raw.charAt(1) !== '-'`

Agree.

`  const inline = token.raw.includes('=')`

Disagree, inline short option and value looks like: `-fFILE`

`  const shortOptionGroup = short && token.raw.length > 2`

Disagree, inline short option and value also triggers this, like `-fFILE`.

`  const shortOption = config.options[token.name]?.short`

Disagree, this returns undefined for short option used without matching `config.options` in `strict: false` mode.
Similar code in `optionUsed` and `shortAndLong` and `indexInShortOptionGroup`.

```
  const indexInShortOptionGroup = shortOptionGroup
    ? token.raw.slice(1).indexOf(config.options[token.name].short)
    : 0
```

Disagree, besides the `shortOption` issue the sub-index is also not so easy. Example: `-fff`.

The token number is simpler for determining ordering than the combination of argument index and group index, so assuming you were trying `indexInShortOptionGroup` from early discussions, may not be needed now.

Reply:
Yeah, in principle it's possible to infer everything from the input `args` array, after all. The whole point of the parser is to reduce the amount of parsing you have to do yourself. So I am in favor of having a little more information.

The example with `strict: false` does make me slightly more in favor of having the actual spelling instead of just a `short` boolean (though I would still prefer that, I think). But I also agree that `raw` is awkward since a token can represent an option plus a value, and that's not how we expect this field to be used. So maybe `rawName`? That is:

```
{
  kind: 'option',
  name: 'verbose',
  rawName: '-v',
  index: 0,
  value: undefined,
}
```

This is _kind of_ a lie because `-xv` would still give you a `rawName` of `-v`, even though that didn't appear in the input (whereas I usually expect `raw` to be something which appears literally in the input). I would prefer just `v`, except that it doesn't allow you to distinguish between `--v` and `-v`, which is unfortunate.

Reply:
Totally agree with the feedback! I'm convinced it makes sense to expose more parse details to authors.

`rawName` is quite clear to me 👍 

In a previous comment, I challenged the `option` prefix in the `token.optionName` detail. However, what are folks thoughts on leaning into an `option` suffix with the `short`/`long` naming convention used throughout the `parseArgs` implementation? e.g.
```js
{
  kind: 'option',
  longOption: 'verbose',
  shortOption: 'v',
  rawOption: '-xv',
  index: 0,
  value: undefined,
  inlineValue: undefined,
}
```

Reply:
> However, what are folks thoughts on leaning into an option suffix with the short/long naming convention used throughout the parseArgs implementation? e.g.

With `shortOption`, either a.) we are unconditionally including `short` even if it was `--long`, in which case the script author needs to parse `raw` to figure out whether it was used as short or long (though I guess that's just checking to see if it starts with `-` or `--`), or b.) we are _not_ unconditionally including `short` for options parsed as `--long`, in which case `shortOption` behaves weirdly different from `longOption`. Neither seems good to me.

(Also, separately, I don't think it's a good idea to include `-xv` anywhere in the `tokens` array. You can get that out of the input `args` if you really want it, but it's basically never useful as-is.)

I don't think we should be trying to treat `short` and `long` as being of equal importance. It's actually pretty unusual to want the short value, for the use cases I've thought about so far - usually you want to do something based off of whether and when a given option was passed, not how it was spelled, and for that purpose the long name is the canonical one.

Reply:
Regarding `rawName`: I wondered whether the "name" in `rawName` implied it didn't have the dash, and had the same "kind of a lie" feeling about it not matching the original text. But coming around to it as I keep looking over the examples!

Reply:
I appreciate the symmetry with the shortOption/longOption/rawOption idea but I agree not of equal importance.

I like `name` as the primary key for looking up input `options` and output `values`.

Reply:
The above examples were considering the previous feedback where we simply expose the parse details we've already derived so authors don't have to.

In regards to `shortOption` being conditionally or unconditionally included, we are already omitting properties based on the input args and don't have any concerns continuing that pattern e.g.
```js
// node cli.js --foo bar -baz -vvv
const {values, tokens} = parseArgs()
tokens === [
  {
    find: 'option',
    long: 'foo',
    short: undefined,// Omitted - This pattern is also used internally when evaluating `config.options`
    raw: '--foo',
    index: 0,
    value: 'bar',
    inlineValue: undefined // Omitted
  }
]
```

Note: I am just voicing my curiosities for the sake of discussion. With that said, what are the downsides to including additional parse details (such as: `short` and `raw`). Thus far, the arguments against have felt a bit opinionated on what we think authors will use most often. I'm not a CLI library author so it's a bit hard for me speak to what information is "critical".

Reply:
The above examples were considering the previous feedback where we simply expose the parse details we've already derived so authors don't have to.

In regards to `shortOption` being conditionally or unconditionally included, we are already omitting token details based on the input args and we don't seem to have issues with this pattern (including myself) e.g.
```js
// node cli.js --foo bar -baz -vvv
const {values, tokens} = parseArgs()
tokens === [
  {
    kind: 'option',
    longOption: 'foo',
    // Omitted - This pattern is also used internally when evaluating `config.options` and
    // additionally becomes a simple mechanism for determining if the input was a `short` or `long` option
    shortOption: undefined,
    rawOption: '--foo',
    index: 0,
    value: 'bar',
    inlineValue: undefined // Omitted - No issues with this pattern
  },
  {
    kind: 'option',
    longOption: 'bar',
    // Here we know a short option was provided to `cli.js`
    shortOption: 'b',
    // Useful for reporting detailed error messages
    rawOption: '-bar',
    index: 1,
    value: undefined, // Omitted - No issues with this pattern
    inlineValue: undefined // Omitted - No issues with this pattern
  },
  ...
  {
    kind: 'option',
    longOption: 'verbose',
    shortOption: 'v',
    // Useful for computing verbosity level e.g. [...token.raw].filter(c => c === 'v').length
    rawOption: '-vvv',
    index: 3,
    value: undefined, // Omitted - No issues with this pattern
    inlineValue: undefined // Omitted - No issues with this pattern
  }
]
```

> I don't think it's a good idea to include `-xv` anywhere in the `tokens` array. You can get that out of the input `args`

This was my original thinking in https://github.com/pkgjs/parseargs/pull/129#discussion_r887445473 and your feedback was: 

> in principle it's possible to infer everything from the input `args` array, after all. The whole point of the parser is to reduce the amount of parsing you have to do yourself

I've since come around and agree with this feedback. I'm thinking these additional parse details almost make it so authors will rarely need to re-evaluate the input `args` (aside from `index` logic).

Note: I am mainly voicing thoughts and curiosities for the sake of discussion. With that said, I'm curious what some downsides would be for including additional parse details (e.g. `longOption`, `shortOption` and `rawOption`)? Some arguments against additional parse details seem to focus on what we think authors will use most often and where they can look if not provided. I'm not a CLI library author so it's a bit hard for me to speak to what information is "critical". Thus, I'm interesting in learning why "including addition parse details X is problematic because Y".

Reply:
> In regards to shortOption being conditionally or unconditionally included, we are already omitting token details based on the input args and we don't seem to have issues with this pattern

It's not that I have a problem with conditionally including things _in general_, it's that I think the naming you've chosen very strongly suggests that `shortOption` and `longOption` are similar kinds of things and so would be expected to behave similarly. But you're proposing that they would not behave similarly, if I've understood correctly: `shortOption` would be conditionally present based on the actual spelling, whereas `longOption` would be unconditionally present. It's specifically having different behavior for similarly-named things which I dislike.

> I've since come around and agree with this feedback. I'm thinking these additional parse details almost make it so authors will rarely need to re-evaluate the input `args` (aside from `index` logic).

The thing I want to emphasize is the "it's basically never useful as-is" part. I don't think we should include anything we don't have a strong use case for. I think there's a strong use case for knowing whether an option was short or not (so you can do custom logic for short options, and so you can give error messages which are customized based on whether the user passed an option as short or long), but I can't think of a common case where you'd want `-xv`.

Basically, I am trying to weigh "how useful would this be if included" and "how hard is it to derive if not included". If something is very useful and at least moderately tricky to derive, it's probably worth including. If it's very hard to derive, maybe it only needs to be only occasionally useful for it to be worth including. But if it's rarely useful _and_ it's pretty easy to derive (for those rare cases), it's not worth including, I think.

Reply:
(Switched from `optionUsed` to `rawName`.)

Reply:
I've added new comments covering some of what was covered here, and will resolve this long thread. Thanks for the thoughtful comments and consideration.